Feat/collapsible if else

lindsaymoralesb opened the pull request:
## Issue
#8 

## Description 

- Lint implementation
- Fix implementation
- Tests added

0xLucqs commented on crates/cairo-lint-core/src/lints/collapsible_if_else.rs:
```diff
@@ -0,0 +1,48 @@
+    let statements: Vec<Statement> = block_expr.statements(db).elements(db);
+
+    // Check if the first statement is an `if` statement
+    if let Some(first_statement) = statements.first() {
+
+        if let Statement::Expr(statement_expr) = first_statement {
+            let first_statement_expr = statement_expr.expr(db);
+            if let Expr::If(_) = first_statement_expr {
+                true
+            } else {
+                false
+            }
+        } else {
+            false
+        }
+    } else {
+        false
+    }
```
too many indents here group the ifs together like that
```suggestion
    if let Some(first_statement) = statements.first() && let Statement::Expr(statement_expr) = first_statement && let Expr::If(_) = first_statement_expr = statement_expr.expr(db) {
    true
    } else {
        false
    }
```

lindsaymoralesb replied:
That suggestion gave some issues, but I improved the code in another way 😊

0xLucqs commented on crates/cairo-lint-core/src/fix.rs:
```diff
@@ -206,6 +210,237 @@ impl Fixer {
+    }
+
+    // Transforms text to replace "else { if" pattern with "else if"
+    fn transform_if_else(&self, text: String) -> String {
```
looks extremely complex, try working with the tree directly not the text. you can do it by replacing the `else` node with the `if` node text and prefix it with `else`

lindsaymoralesb replied:
got it, working on it!

lindsaymoralesb commented:
hey @0xLucqs comments addressed!

1. For the nested ifs change on the linter your suggestion was generating some compilation issues, so I fixed it in another way.
2. For the complex fix, I changed the approach I was using (with the text) and instead started using the tree. This is working good, I just noticed that there is a small indentation issue with the fix on the else if clause. Do you have an idea of how can I approach that small issue?

0xLucqs commented:
let's ignore complex indentation problems it's a nightmare

lindsaymoralesb commented:
> let's ignore complex indentation problems it's a nightmare

Got it! Then the PR is ready for review.

lindsaymoralesb commented:
@0xLucqs is there any action needed from my side to make the ci/cd run correctly?

0xLucqs commented:
Yep you need to run `cargo fmt`

lindsaymoralesb commented:
> Yep you need to run `cargo fmt`

done! @0xLucqs

0xLucqs commented:
@all-contributors please add @lindsaymoralesb for code and test
